### PR TITLE
[ART-2682] Associate builds with CVE bugs

### PR DESCRIPTION
**Pull request by vfreex**
- attach-cve-flaws: associate builds with CVE bugs after flaw bugs are attached
- verify-attached-bugs: verify that CVE bugs are correctly associate attached builds

### CODE REVIEW

File: elliottlib/errata_async.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    async def _make_request(self, method: str, path: str, parse_json: bool = True, **kwargs) -> Union[Dict, bytes]:
+        if "headers" not in kwargs:
+            kwargs["headers"] = self._headers
+        async with self._session.request(method, self._errata_url + path, **kwargs) as resp:
+            resp.raise_for_status()
+            result = await (resp.json() if parse_json else resp.read())
+        return result
+
+    async def get_advisory(self, advisory: Union[int, str]) -> Dict:
+        path = f"/api/v1/erratum/{quote(str(advisory))}"
```
Comment (joepvd) [inline elliottlib/errata_async.py:45]: Could paths like these be consumed from `elliottlib/constants.py`?

File: elliottlib/cli/attach_cve_flaws_cli.py [hunk 2]
```diff
@@ -105,7 +117,23 @@ def attach_cve_flaws_cli(runtime, advisory_id, noop, default_advisory_type):
     if noop:
         return True
 
-    return advisory.commit()
+    advisory.commit()
+
+    # Associate builds with CVEs
```
Comment (joepvd) [inline elliottlib/cli/attach_cve_flaws_cli.py:122]: Could this be in a dedicated method, and put under test?

File: elliottlib/errata_async.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import aiohttp
+import gssapi
+from kobo.rpmlib import parse_nvr
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class AsyncErrataAPI:
+    def __init__(self, url: str):
+        self._errata_url = urlparse(url).geturl()
```
Comment (joepvd) [inline elliottlib/errata_async.py:16]: Any reason constans.py is not consumed here?

File: elliottlib/errata_async.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        _LOGGER.debug("Creating current CVE exclusions for advisory %s", advisory_id)
+        current_cve_package_exclusions = {}
+        async for cve_package_exclusion in api.get_cve_package_exclusions(advisory_id):
+            current_cve_package_exclusions.setdefault(cve_package_exclusion["relationships"]["cve"]["name"], {})[cve_package_exclusion["relationships"]["package"]["name"]] = cve_package_exclusion["id"]
+
+        extra_cve_package_exclusions = {cve: exclusions.keys() - expected_cve_packages_exclusions.get(cve, set()) for cve, exclusions in current_cve_package_exclusions.items()}
+        missing_cve_package_exclusions = {cve: exclusions - current_cve_package_exclusions.get(cve, {}).keys() for cve, exclusions in expected_cve_packages_exclusions.items()}
+
+        _LOGGER.debug("Reconsiling CVE package exclusions for advisory %s", advisory_id)
```
Comment (joepvd) [inline elliottlib/errata_async.py:105]: Could the loglevel for application logs be `info`?
Part of the reason why I think that'd be nice is to not accidentally get a debug level logging for http requests, which might leak kerberos tokens.

Comment (joepvd) [reply elliottlib/errata_async.py:105]: Also, `s/Reconsiling/Reconciling/` :)

File: elliottlib/errata_async.py (hunk 3 above)
Comment (vfreex) [reply elliottlib/errata_async.py:16]: A client instantiating AsyncErrataAPI usually get Errata URL from erratatool.yml in ocp-build-data. I don't want to couple things unrelated to API calls here.

File: elliottlib/cli/attach_cve_flaws_cli.py (hunk 2 above)
Comment (vfreex) [reply elliottlib/cli/attach_cve_flaws_cli.py:120]: Good idea. I haven't added any unit tests yet. This card is more complex than I estimated :(
Will add unit tests in the next push.

File: elliottlib/cli/attach_cve_flaws_cli.py [hunk 5]
```diff
@@ -100,12 +112,32 @@ def attach_cve_flaws_cli(runtime, advisory_id, noop, default_advisory_type):
...
+    try:
+        await errata_api.login()
+        await associate_builds_with_cves(errata_api, advisory, attached_tracker_bugs, tracker_flaws, flaw_id_bugs, noop)
+    finally:
+        await errata_api.close()
+
+
+async def associate_builds_with_cves(errata_api: AsyncErrataAPI, advisory: Erratum, attached_tracker_bugs: List[Bug], tracker_flaws: Dict[int, List[int]], flaw_id_bugs: Dict[int, Bug], dry_run: bool):
+    attached_builds = [b for pv in advisory.errata_builds.values() for b in pv]
+    rpm_cves: Dict[Bug, str] = bzutil.get_valid_rpm_cves(attached_tracker_bugs)
```
Comment (jasinner) [inline elliottlib/cli/attach_cve_flaws_cli.py:133]: Why does this exclude container cves? Does this feature work with CVE for containers?

**Comment by vfreex** [comment]
@jasinner I've updated this PR. Now it should work with containers. PTAL.

**Comment by vfreex** [comment]
PR rebased

**Comment by vfreex** [comment]
Errata API "/cve/show/{advisory}.json" doesn't return the correct CVEs for some RHSAs.
 Not sure if it's an Errata bug. Updated this PR to use a different approach instead.

**Comment by vfreex** [comment]
PR rebased.

**Comment by joepvd** [comment]
:shipit:

File: elliottlib/errata_async.py (hunk 1 above)
Comment (locriandev) [reply elliottlib/errata_async.py:45]: The API endpoint string is used in more than one place, how about moving its definition at module level?

**Pull request merged**